### PR TITLE
KAS-FFC

### PR DESCRIPTION
**Pull request by bfussell**
registration, vector parsing and responses

### CODE REVIEW

File: app/app_main.c [hunk 1]
```diff
@@ -3719,6 +3765,114 @@ static ACVP_RESULT app_kas_ecc_handler(ACVP_TEST_CASE *test_case)
...
+    if (Z) {
+        OPENSSL_cleanse(Z, Zlen);
+    }
+    OPENSSL_free(Z);
+    BN_clear_free(peerkey);
+    BN_clear_free(dh->priv_key);
+    BN_clear_free(dh->pub_key);
+    dh->priv_key = NULL;
+    dh->pub_key = NULL;
+	FIPS_dh_free(dh);
```
Comment (fliphil) [inline app/app_main.c]: Nit, too much space

File: src/acvp_kas_ffc.c [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+    memset(tmp, 0x0, ACVP_KAS_FFC_MAX_STR);
+    rv = acvp_bin_to_hexstr((const unsigned char *)stc->piut, stc->piutlen, 
+                            (unsigned char *) tmp);
+    if (rv != ACVP_SUCCESS) {
+        free(tmp);
+        ACVP_LOG_ERR("hex conversion failure (Z)");
+        return rv;
+    }
+    json_object_set_string(tc_rsp, "ephemeralPublicIut", stc->piut);
```
Comment (fliphil) [inline src/acvp_kas_ffc.c]: Should you be passing in tmp here instead of stc->piut?

File: src/acvp_kas_ffc.c [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+            rv = acvp_kas_ffc_init_comp_tc(ctx, stc, tc_id, hash,
+                                           p, q, g, eps, epri, epui, z, mode);
+            if (rv != ACVP_SUCCESS) {
+                return ACVP_CRYPTO_MODULE_FAIL;
+            }
+
+            /* Process the current KAT test vector... */
+            rv = (cap->crypto_handler)(tc);
+            if (rv != ACVP_SUCCESS) {
+                    return ACVP_CRYPTO_MODULE_FAIL;
```
Comment (fliphil) [inline src/acvp_kas_ffc.c]: Nit too much space

File: src/acvp_kas_ffc.c [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+            ACVP_LOG_ERR("ACVP server requesting unsupported capability");
+            return ACVP_UNSUPPORTED_OP;
+        }
+        rv = acvp_kas_ffc_comp(ctx, cap, &tc, &stc, obj, mode, r_tarr);
+        break;        
+    case ACVP_KAS_FFC_MODE_NOCOMP:
+    default:
+        ACVP_LOG_ERR("ACVP server requesting unsupported KAS-FFC mode");
+        return ACVP_UNSUPPORTED_OP;
+        break;
```
Comment (fliphil) [inline src/acvp_kas_ffc.c]: Nit, break not needed after return

File: src/acvp.h [hunk 5]
```diff
@@ -1608,6 +1698,84 @@ ACVP_RESULT acvp_enable_kas_ecc_cap_scheme (ACVP_CTX *ctx,
...
+
+    @return ACVP_RESULT
+ */
+ACVP_RESULT acvp_enable_kas_ffc_cap_parm (ACVP_CTX *ctx,
+                                          ACVP_CIPHER cipher,
+                                          ACVP_KAS_FFC_MODE mode,
+                                          ACVP_KAS_FFC_PARAM param,
+                                          int value);
+
+ACVP_RESULT acvp_enable_kas_ffc_cap_scheme (ACVP_CTX *ctx,
```
Comment (fliphil) [inline src/acvp.h]: Doxygen comment for this function?

File: app/app_main.c (hunk 1 above)
Comment (bfussell) [reply app/app_main.c]: yep, one of those nasty tabs got left in there.

File: src/acvp_kas_ffc.c (hunk 4 above)
Comment (bfussell) [reply src/acvp_kas_ffc.c]: fixed

File: src/acvp.h (hunk 5 above)
Comment (bfussell) [reply src/acvp.h]: fixed

**Pull request merged**